### PR TITLE
'apply_resolutions' reports error at the end if any resolution failed

### PR DESCRIPTION
**Pull request by beyhan**
- Before this change, 'apply_resolutions' always succeeded.
- This changes behaviour of job 'scan_and_fix' and 'apply_resolutions'.
  They will finish with error now in case of any resolution error.

[#114092983](https://www.pivotaltracker.com/story/show/114092983)

Signed-off-by: Tom Kiemes <tom.kiemes@sap.com>

### CODE REVIEW

**Comment by beyhan** [comment]
Problem resolver task has error state and looks like:

```
, [2016-02-25 14:05:53 #13178] [task:55] ERROR -- DirectorJobRunner: Error resolving problem `6': Timed out pinging to c31b4b86-2e7d-4282-b1ba-26a55c4cc348 after 600 seconds
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/problem_resolver.rb:48:in `apply_resolutions'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/jobs/cloud_check/scan_and_fix.rb:37:in `block in perform'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/lock_helper.rb:13:in `block in with_deployment_lock'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/lock.rb:56:in `lock'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/lock_helper.rb:13:in `with_deployment_lock'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/jobs/cloud_check/scan_and_fix.rb:30:in `perform'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/job_runner.rb:99:in `perform_job'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/job_runner.rb:32:in `block in run'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh_common-1.0000.0/lib/common/thread_formatter.rb:49:in `with_thread_name'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/job_runner.rb:32:in `run'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/lib/bosh/director/jobs/base_job.rb:10:in `perform'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/resque-1.25.2/lib/resque/job.rb:227:in `perform'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/resque-1.25.2/lib/resque/worker.rb:250:in `perform'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/resque-1.25.2/lib/resque/worker.rb:189:in `block in work'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/resque-1.25.2/lib/resque/worker.rb:166:in `loop'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/resque-1.25.2/lib/resque/worker.rb:166:in `work'
/var/vcap/packages/director/gem_home/ruby/2.1.0/gems/bosh-director-1.0000.0/bin/bosh-director-worker:75:in `<top (required)>'
/var/vcap/packages/director/bin/bosh-director-worker:16:in `load'
/var/vcap/packages/director/bin/bosh-director-worker:16:in `<main>'
D, [2016-02-25 14:05:53 #13178] [task:55] DEBUG -- DirectorJobRunner: (0.000478s) SELECT NULL
D, [2016-02-25 14:05:53 #13178] [task:55] DEBUG -- DirectorJobRunner: (0.000136s) BEGIN
D, [2016-02-25 14:05:53 #13178] [task:55] DEBUG -- DirectorJobRunner: (0.000690s) UPDATE "tasks" SET "state" = 'error', "timestamp" = '2016-02-25 14:05:53.131267+0000', "description" = 'scan and fix', "result" = 'Error resolving problem `6'': Timed out pinging to c31b4b86-2e7d-4282-b1ba-26a55c4cc348 after 600 seconds', "output" = '/var/vcap/store/director/tasks/55', "checkpoint_time" = '2016-02-25 13:55:00.631683+0000', "type" = 'cck_scan_and_fix', "username" = 'hm' WHERE ("id" = 55)
D, [2016-02-25 14:05:53 #13178] [task:55] DEBUG -- DirectorJobRunner: (0.002356s) COMMIT
I, [2016-02-25 14:05:53 #13178] []  INFO -- DirectorJobRunner: Task took 10 minutes 52.512512077 seconds to process.

Task 55 error
```
`bosh tasks recent --no-filter` shows:
```
+----+------------+-------------------------+-------+-------------------+--------------------------------------------------------------------------------+
| #  | State      | Timestamp               | User  | Description       | Result                                                                         |
+----+------------+-------------------------+-------+-------------------+--------------------------------------------------------------------------------+
| 65 | error      | 2016-02-25 14:16:51 UTC | hm    | scan and fix      | Error resolving problem `7': Timed out pinging to...                           |
| 69 | error      | 2016-02-25 14:09:59 UTC | hm    | scan and fix      | Unable to get deployment lock, maybe a deployment is in progress. Try again... |
| 68 | error      | 2016-02-25 14:08:59 UTC | hm    | scan and fix      | Unable to get deployment lock, maybe a deployment is in progress. Try again... |
| 66 | error      | 2016-02-25 14:06:59 UTC | hm    | scan and fix      | Unable to get deployment lock, maybe a deployment is in progress. Try again... |
| 55 | error      | 2016-02-25 14:05:53 UTC | hm    | scan and fix      | Error resolving problem `6': Timed out pinging to...                           |
```

File: bosh-director/spec/unit/jobs/cloud_check/apply_resolutions_spec.rb [hunk 1]
```diff
@@ -20,28 +20,44 @@ module Bosh::Director
...
 
-    it 'should normalize the problem ids' do
-      allow(job).to receive(:with_deployment_lock).and_yield
+    describe '#perform' do
+      context 'when resolution succeeds' do
+        it 'should normalize the problem ids' do
+          allow(job).to receive(:with_deployment_lock).and_yield
 
-      expect(resolver).to receive(:apply_resolutions).with(normalized_resolutions)
+          expect(resolver).to receive(:apply_resolutions).with(normalized_resolutions)
```
Comment (voelzmo) [inline bosh-director/spec/unit/jobs/cloud_check/apply_resolutions_spec.rb]: funny whitespace?

File: bosh-director/spec/unit/problem_resolver_spec.rb [hunk 2]
```diff
@@ -21,76 +21,102 @@ def inactive_disk(id, deployment_id = nil)
...
-      problems = []
+    describe '#apply_resolutions' do
+      context 'when execution succeeds' do
+        it 'applies all resolutions' do
+          disks = []
+          problems = []
 
-      agent = double('agent')
-      expect(agent).to receive(:list_disk).and_return([])
+          agent = double('agent')
```
Comment (voelzmo) [inline bosh-director/spec/unit/problem_resolver_spec.rb]: let's fix the whitespace here

**Comment by beyhan** [comment]
I created a new one against develop branch.

**Comment by voelzmo** [comment]
New PR against the `develop` branch: https://github.com/cloudfoundry/bosh/pull/1152

**Pull request closed without merging**